core/storage/statistics/gae_models.py

```diff
@@ -59,6 +59,183 @@ def get_or_create(cls, exploration_id, state_name):
+    client_time_spent_in_secs = ndb.FloatProperty(indexed=True)
+    # Whether the submitted answer received useful feedback
+    is_feedback_useful = ndb.BooleanProperty(indexed=True)
+    # Whether the submitted answer is the correct answer.
+    is_answer_correct = ndb.BooleanProperty(indexed=True)
+
+    @classmethod
+    def get_new_event_entity_id(cls, exp_id, session_id):
+        timestamp = datetime.datetime.utcnow()
+        return cls.get_new_id('%s:%s:%s' % (
+            utils.get_time_in_millisecs(timestamp),
+            exp_id,
+            session_id))
+
+    @classmethod
+    def create(cls, exp_id, exp_version, state_name, session_id,
+               client_time_spent_in_secs, is_feedback_useful,
+               is_answer_correct):
+        """Creates a new answer submitted event."""
+        entity_id = cls.get_new_event_entity_id(
+            exp_id, session_id)
+        answer_submitted_event_entity = cls(
+            id=entity_id,
+            event_type=feconf.EVENT_TYPE_ANSWER_SUBMITTED,
+            exp_id=exp_id,
+            exp_version=exp_version,
+            state_name=state_name,
+            session_id=session_id,
+            client_time_spent_in_secs=client_time_spent_in_secs,
+            is_feedback_useful=is_feedback_useful,
+            is_answer_correct=is_answer_correct)
+        answer_submitted_event_entity.put()
+        return entity_id
+
+
+class ExplorationActualStartEventLogEntryModel(base_models.BaseModel):
+    """An event triggered by a student entering an exploration. In this context,
+    'actually' entering an exploration means the student has spent a defined
+    MIN_TIME on the exploration.
+
+    Event schema documentation
+    --------------------------
+    V1:
+        event_type: 'actual_start'
+        exp_id: id of exploration currently being played
+        exp_version: version of exploration
+        state_name: Name of current state
+        client_time_spent_in_secs: Time since start of the exploration when
+            this event was recorded (MIN_TIME to record this event)
+        session_id: ID of current student's session
+    """
+    # Which specific type of event this is
+    event_type = ndb.StringProperty(indexed=True)
+    # Id of exploration currently being played.
+    exp_id = ndb.StringProperty(indexed=True)
+    # Current version of exploration.
+    exp_version = ndb.IntegerProperty(indexed=True)
+    # Name of current state.
+    state_name = ndb.StringProperty(indexed=True)
+    # ID of current student's session
+    session_id = ndb.StringProperty(indexed=True)
+    # Time since start of this state before this event occurred (in sec). This
+    # should be equal to the MIN_TIME defined for recording this event.
+    client_time_spent_in_secs = ndb.FloatProperty(indexed=True)
+
+    @classmethod
+    def get_new_event_entity_id(cls, exp_id, session_id):
+        timestamp = datetime.datetime.utcnow()
```

Comment:
ditto. It will be good to explain id format.

Reply:
Done.